### ISSUE TITLE
All UWP packages are broken on .NET 9 because of Uno references

### ISSUE THREAD

**Issue by Sergio0694** [CommunityToolkit/Tooling-Windows-Submodule#238]
### Describe the bug

I was testing the new builds from `main` and noticed all UWP packages are broken on .NET 9:

![Image](https://github.com/user-attachments/assets/a14f57b0-282f-4f61-8e0c-cabdfdb0b0f1)

The core of the issue is that we have a `net9.0` TFM pulling in Uno dependencies. We should not do this. The official guidance for packages that have UI framework dependencies is to have **different packages** for different UI frameworks. Uno should have their own packages as well. We need 4:
- `*.Uwp`: UWP .NET Native and .NET 9
- `*.Uwp.Uno`: UWP XAML Uno flavor
- `*.WinUI`: WinUI 3
- `*.WinUI.Uno`: WinUI 3 Uno flavor

As a **short term solution**, we can add a `net9.0-windows10.0.26100.0` TFM to work around the issue for UWP .NET 9 consumers.

### Steps to reproduce

```text
n/a
```

### Expected behavior

n/a

### Code Platform

- [x] UWP
- [ ] WinAppSDK / WinUI 3
- [ ] Web Assembly (WASM)
- [ ] Android
- [ ] iOS
- [ ] MacOS
- [ ] Linux / GTK

### Windows Build Number

- [ ] Windows 10 1809 (Build 17763)
- [ ] Windows 10 1903 (Build 18362)
- [ ] Windows 10 1909 (Build 18363)
- [ ] Windows 10 2004 (Build 19041)
- [ ] Windows 10 20H2 (Build 19042)
- [ ] Windows 10 21H1 (Build 19043)
- [ ] Windows 11 21H2 (Build 22000)
- [ ] Other (specify)

### App minimum and target SDK version

- [ ] Windows 10, version 1809 (Build 17763)
- [ ] Windows 10, version 1903 (Build 18362)
- [ ] Windows 10, version 1909 (Build 18363)
- [ ] Windows 10, version 2004 (Build 19041)
- [ ] Other (specify)

### Help us help you

Yes, I'd like to be assigned to work on this item.

**Comment by Arlodotexe** [comment]
# Background
Packaging was first introduced in https://github.com/CommunityToolkit/Labs-Windows/issues/28 (see [here](https://github.com/CommunityToolkit/Labs-Windows/pull/75/files#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721R65)). At the time, our infrastructure was capable of multi-targeting but _not packaging_, primarily because we didn't have a way to distinguish between the two variants of Uno (Uno.UI/Uwp and Uno.WinUI/Wasdk) either while building or while packaging, since they used the same `TargetFramework` on both.

If not for Uno here, we likely would have put Uwp and Wasdk into one package, and we wouldn't have been able to [ship .NET 9 in UWP](https://github.com/CommunityToolkit/Tooling-Windows-Submodule/pull/235) alongside Wasdk like we just did, since the Windows App SDK and modern .NET on UWP use the same TFM.

@michael-hawker and I were under the impression that NuGet and the .NET SDK would be enough to handle what we were doing, and that we'd run into total blockers if we did something we shouldn't. We never got that signal beyond the Uwp/WinUI package split needed to support Uno in the first place, which happened to work in our favor in regard to official guidance.  

We also encountered this issue in https://github.com/CommunityToolkit/Tooling-Windows-Submodule/pull/125, where we also found a tool that allowed us to check TFM compatibility and precedence:
> Using [this](https://nugettools.azurewebsites.net/) tool, you can see that:
> - For tfm compatibility (see [here](https://nugettools.azurewebsites.net/6.7.0/framework-compatibility?project=net7.0-windows10.0.19041.0&package=net6.0-windows10.0.19041.0)), `net7.0-windows10.0.19041.0` projects _can_ use `net6.0-windows10.0.19041.0` packages
> - For tfm precedence (see [here](https://nugettools.azurewebsites.net/6.7.0/framework-precedence?framework=net7.0-windows10.0.19041.0&exludedIdentifiers=)), the very next candidate after `net7.0-windows10.0.19041.0` is `net7.0`.

# Problem

The problem here seems to be that TFM precedence is affecting our ability to use the `net8.0-windows*` TFM from UWP projects targeting `net9.0-windows*`, because the package contains a `net9.0` TFM which dotnet selects instead of selecting `net8.0-windows*` like we expect it might. 

# Solution

Revisit official guidance. 

@Sergio0694 had stated:
> The official guidance for packages that have UI framework dependencies is to have different packages for different UI frameworks.

Checking around the web, I've found:
--

but I'm unable to find any documentation to support this.

**Comment by Sergio0694** [comment]
This general issue came up while we were evaluating what TFM/TPM we should use for UWP support for .NET 9. We eventually settled on just using `netX.0-windows10...` like all other UI frameworks (Windows Forms, WPF, WinUI, etc.) are also doing (it was UWP that was the odd one out). While doing this, the question of how to handle packages targeting multiple UI frameworks was also raised. And once again the official answer is just: **have different packages for each specific UI framework**.

Also worth mentioning that the whole TFM <=> UI framework association would never be correct anyway. It's completely valid to mix and match multiple UI frameworks in the same project too. That's another reason why having multiple packages is the only way to give developers proper control over what they're pulling in.

**Comment by Arlodotexe** [comment]
@Sergio0694 Just to clarify:
- UWP XAML and WinUI 3 are separate UI frameworks
- Uno.UI is a 1:1 match of the APIs in UWP XAML
- Uno.WinUI is a 1:1 match of many APIs in the Windows App SDK.

To understand this issue properly, we need to recognize that Uno designed to be a 'separate UI framework', rather it's specifically design to polyfill UWP XAML and WinUI 3 as transparently as possible. If it runs natively, it should 'just work' on the other platforms.

If we were instead some hypothetical official implementations of WinUI on those platforms instead of using Uno, _**it would still have this TFM precedence issue.**_. 

That's important to note, because the issue can be solved a different way: by using _more specific TFMs_ instead of simply `net9.0`, such as `net9.0-wasm`, just as MAUI would do to work around the issue.

**Comment by Arlodotexe** [comment]
> If we were instead using some hypothetical official implementations of WinUI on those platforms instead of using Uno, _**it would still have this TFM precedence issue**_. 
> 
> That's important to note, because the issue can be solved a different way: by using _more specific TFMs_ instead of simply `net9.0`, such as `net9.0-wasm`, just as MAUI would do to work around the issue. 

@jeromelaban Is this something that's been raised with the Uno Platform team? Do you have any insights to add from your end?

**Comment by Sergio0694** [comment]
> "If we were instead using some hypothetical official implementations of WinUI on those platforms instead of using Uno, it would still have this TFM precedence issue."

Yeah, the problem isn't specific to Uno. We're just only hitting it with Uno because UWP and WinAppSDK are already split.

> "especially since we still can't find published guidance on this"

Like I said, the official guidance is to split libraries into separate packages for different UI frameworks (WPF, UWP, WinUI, etc.). I agree there doesn't seem to be a dedicated docs page specifically for this. We talked about adding a couple bullet points for this to [this docs page](https://learn.microsoft.com/en-us/nuget/create-packages/package-authoring-best-practices), but we haven't done so just yet (the guidance is however what I've mentioned).

I'll make a PR into the docs repo to fix this 🙂

> "by using more specific TFMs instead of simply `net9.0`"

Right, the issue would also go away if you removed **all** `-windows` TFMs for Uno. But I don't think that's viable?

**Comment by Arlodotexe** [comment]
> Yeah, the problem isn't specific to Uno. We're just only hitting it with Uno because UWP and WinAppSDK are already split.

Close, but not quite. We would have hit the same problem is we had gotten our "unified package" like we initially wanted, but in a different way.

If we had supported all four of Uwp, Wasdk, and Uwp/Uno and Wasdk/Uno (e.g., if Uno used different TFMs per platform from the start), then the split could have occurred either down Uwp/Wasdk or Uno/Non-Uno first. One split, regardless of which we did first, wouldn't have been enough. The existing UWP and WinAppSDK package split was only ever done because of Uno. 

> Right, the issue would also go away if you removed all -windows TFMs for Uno. But I don't think that's viable?

No, that's not quite right either. Uno _isn't using_ any -windows TFMs. The issue runs the other way around: we need a _more specific_ TFM for each platform than `net9.0` to avoid the TFM precedence issues.

**Comment by Sergio0694** [comment]
> "instead of flying blind making decisions"

Just to clarify, we're not flying blind here. I'm telling you what the official guidance is 🙂
And that's from several internal conversations we've had with folks from both .NET, NuGet, and Windows. I agree that we should also have this guidance being written down somewhere in the docs (and I'll help with that), but we're not flying blind here.

> "TFM to incorrectly select when running a higher and more platforms-specific TFM than our package supports"

Note that having more specific TFMs wouldn't fully solve the problem either. The fundamental issue is that you _cannot_ (by design) rely on TFMs to know which UI framework someone is consuming. Nothing stops people from referencing _more than a single UI framework_ in a given project (eg. you could have a WindowsAppSDK app with a UWP XAML island, or you could have a WPF app that also uses WindowsAppSDK stuff, and so on). So having separate packages is also the only way to let people properly control which dependencies they're pulling in, based on what they need and want to do.

**Comment by Arlodotexe** [comment]
> Just to clarify, we're not flying blind here. I'm telling you what the official guidance is 🙂

Right, to clarify: We're flying blind, you're not. You're telling me the what but none of the why, which doesn't allow us to inspect all of our options or identify possible alternatives. That's why having guidance and underlying relational documented would be nice. 

> The fundamental issue is that you cannot (by design) rely on TFMs to know which UI framework someone is consuming. 

Right... as I laid out, Uno isn't a _different_ UI framework, it's a re-implementation of UWP XAML and WinUI 3 on other TFMs to run on other platforms.

The problem is that TFM precedence will prioritize a non-platform-specific TFM if the version is a closer: `net9.0` gets picked instead of `net8.0-windows*` by a project running `net9.0-windows*`. 

This wouldn't happen if it used `net8.0-wasm` or `net9.0-wasm`, because that TFM isn't compatible with `netX.0-windows*`, so `net9.0-windows*` projects would have grabbed `net8.0-windows*` from the package instead.

---

All that said, you're saying something slightly different @Sergio0694. We should stress both. 

To me, it sounds like you're saying that each _UI implementation_ (including polyfills like Uno) should be package-specific, not just the targeted _UI framework_ itself (e.g. UWP XAML vs WinUI 3). Is that right?

That is, something like MAUI which uses the _same underlying UI implementation_ for all targets would be able to package together in the same NuGet package, but something like Uno that uses a _different implementation_ from some UI targets should be in a separate package?

**Comment by Sergio0694** [comment]
> "To me, it sounds like you're saying that each UI implementation (including polyfills like Uno) should be package-specific, not just the targeted UI framework itself (e.g. UWP XAML vs WinUI 3). Is that right?"

Yes, that's what I'm saying and why I feel there's some disconnect. The fact that Uno is technically a reimplementation of whatever UI framework and not a UI framework per se doesn't matter. The point is that you should not use TFMs to select which technology to pull in. You should have different NuGet packages _per technology_, and then each will just have TFMs for whichever target framework they can support.

**Comment by Arlodotexe** [comment]
@Sergio0694 Perfect, that makes sense! Glad we could get it cleared up.

**Comment by jeromelaban** [comment]
The platform neutral support of Uno has one purpose at this time: Provide a "reference" implementation of Uno APIs. Uno uses this to allow for libraries that don't need to have platform specific code to be available on WebAssembly, desktop and other targets that don't need platform specific APIs. This allows the use of UI (WinUI 3 APIs) and non-UI APIs (WinRT APIs such as File Pickers, `Windows.Storage`, Sensors, etc...) without having to multi-target.

About the platform neutral fallback, that is a behavior of NuGet which is a side effect of a choice we made a very long time ago when `netstandard2.0` was in prime time, and it has stayed that way since then. We needed to have a TFM that would work with WebAssembly, and it was the only one available at the time. The fix for now is to add a "placeholder" target framework that duplicates the binaries in order to avoid the fallback.

> @Sergio0694 You should have different NuGet packages per technology, and then each will just have TFMs for whichever target framework they can support.

We have a difference of opinion here. TFMs are implicitly specifying a technology, when having their platform part specified which brings implicitly underlying dependencies. Specifically `netX.0-windows10.yyy` implies Win32/WPF, which is a UI technology. Same goes for iOS/Android.

> You should have different NuGet packages per technology, and then each will just have TFMs for whichever target framework they can support.

Specifically for the Windows Community Toolkit, Uno implements the same APIs as the WinAppSDK counterpart which makes it for a natural exclusion of windows that is "owned" by Win32/WinUI/WPF. This makes for the obvious choice of making all the other targets framework available to support the WCT features for non-Windows targets. The resulting packages effectively support all of the required target frameworks.

This is a package authoring choice to make packages consumption easy, where the NuGet package manager does not help developers at all to create `PackageReference` conditionals.

That being said, creating multiple packages that target individual UI technologies is something we've been doing since the beginning of Uno 7 years ago, it's definitely doable yet simply not user friendly.

**Comment by michael-hawker** [comment]
I definitely think we should push NuGet here to do better and allow being able to have dlls and components for multiple technologies (WPF, WinUI, Uno, etc...) that support multiple platforms (WASM, iOS, Android, Windows). This really is a crux of the problem.

As a library maintainer/author, shipping multiple packages is a PITA. As a consumer, figuring out which package to use is error prone and a PITA. Especially as the packages don't know what tech they really support the tooling can't always say it's the wrong for the user that they have the wrong one.

It'd be great to look at this from the customers/developers point of view, both as a developer building an app, and a library author in the .NET ecosystem. Can we work with NuGet to make this better for everyone? This just gets more complex year-over-year as .NET is supported and used in more technologies across a greater number of platforms.

Current Related NuGet Issues:

- https://github.com/NuGet/Home/issues/11645
- https://github.com/NuGet/Home/issues/7416
- https://github.com/NuGet/Home/issues/8186

I feel like what we're talking about is a cross-section of all these above issues. Do we need an overarching better issue to be filed in their repo to have start/have this discussion?

**Comment by Sergio0694** [comment]
> "We have a difference of opinion here. TFMs are implicitly specifying a technology, when having their platform part specified which brings implicitly underlying dependencies. Specifically `netX.0-windows10.yyy` implies Win32/WPF, which is a UI technology. Same goes for iOS/Android."

Forgot to reply to this. This is just not true though. Using `netX.0-windows10.0.YYYY` does **not** imply WPF at all. Case in point, it's the same TFM that UWP and WinUI 3 also use. And there's absolutely 0 references to WPF being added by default.
